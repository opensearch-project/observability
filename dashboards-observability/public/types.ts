/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { DashboardStart } from '../../../src/plugins/dashboard/public';
import { NavigationPublicPluginStart } from '../../../src/plugins/navigation/public';

export interface AppPluginStartDependencies {
  navigation: NavigationPublicPluginStart;
  dashboard: DashboardStart;
}

export interface ObservabilitySetup {}

export interface ObservabilityStart {}
