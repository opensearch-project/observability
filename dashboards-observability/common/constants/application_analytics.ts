/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export const TAB_OVERVIEW_ID = 'app-analytics-overview';
export const TAB_SERVICE_ID = 'app-analytics-service';
export const TAB_TRACE_ID = 'app-analytics-trace';
export const TAB_LOG_ID = 'app-analytics-log';
export const TAB_PANEL_ID = 'app-analytics-panel';
export const TAB_CONFIG_ID = 'app-analytics-config';
export const TAB_OVERVIEW_TITLE = 'Overview';
export const TAB_SERVICE_TITLE = 'Services';
export const TAB_TRACE_TITLE = 'Traces & Spans';
export const TAB_LOG_TITLE = 'Log Events';
export const TAB_PANEL_TITLE = 'Panel';
export const TAB_CONFIG_TITLE = 'Configuration';

export const APP_ANALYTICS_API_PREFIX = '/api/observability/application';
