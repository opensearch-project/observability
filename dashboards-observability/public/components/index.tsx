/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import ReactDOM from 'react-dom';
import { QueryManager } from 'common/query_manager';
import { AppMountParameters, CoreStart } from '../../../../src/core/public';
import { AppPluginStartDependencies } from '../types';
import { App } from './app';

export const Observability = (
  CoreStartProp: CoreStart,
  DepsStart: AppPluginStartDependencies,
  AppMountParametersProp: AppMountParameters,
  pplService: any,
  dslService: any,
  savedObjects: any,
  timestampUtils: any,
  queryManager: QueryManager
) => {
  ReactDOM.render(
    <App
      CoreStartProp={CoreStartProp}
      DepsStart={DepsStart}
      pplService={pplService}
      dslService={dslService}
      savedObjects={savedObjects}
      timestampUtils={timestampUtils}
      queryManager={queryManager}
    />,
    AppMountParametersProp.element
  );

  return () => ReactDOM.unmountComponentAtNode(AppMountParametersProp.element);
};
