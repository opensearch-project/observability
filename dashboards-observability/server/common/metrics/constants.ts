/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { ComponentType, CounterType } from './types';

export const WINDOW = 3600;
export const INTERVAL = 60;
export const CAPACITY = (WINDOW / INTERVAL) * 2;
export const MILLIS_MULTIPLIER = 1000;

export const COMPONENTS = [
  'application_analytics',
  'operational_panels',
  'event_analytics',
  'notebooks',
  'trace_analytics',
  'metrics_analytics',
] as const;
export const REQUESTS = ['create', 'get', 'update', 'delete'] as const;

export const GLOBAL_BASIC_COUNTER: CounterType = (() => {
  const counter = {} as CounterType;
  COMPONENTS.forEach((component) => {
    counter[component] = {} as CounterType[ComponentType];
    REQUESTS.forEach((request) => {
      counter[component][request] = {
        total: 0,
      };
    });
  });
  return counter;
})();

export const DEFAULT_ROLLING_COUNTER: CounterType = (() => {
  const counter = {} as CounterType;
  COMPONENTS.forEach((component) => {
    counter[component] = {} as CounterType[ComponentType];
    REQUESTS.forEach((request) => {
      counter[component][request] = {
        count: 0,
        system_error: 0,
        user_error: 0,
      };
    });
  });
  return counter;
})();
