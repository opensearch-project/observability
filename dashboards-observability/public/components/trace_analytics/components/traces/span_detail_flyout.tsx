/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  EuiButtonIcon,
  EuiCodeBlock,
  EuiCopy,
  EuiFlexGroup,
  EuiFlexItem,
  EuiFlyout,
  EuiFlyoutBody,
  EuiFlyoutHeader,
  EuiHorizontalRule,
  EuiSpacer,
  EuiText,
  EuiTitle,
} from '@elastic/eui';
import _ from 'lodash';
import moment from 'moment';
import React, { useEffect, useState } from 'react';
import { HttpSetup } from '../../../../../../../src/core/public';
import { TRACE_ANALYTICS_DATE_FORMAT } from '../../../../../common/constants/trace_analytics';
import { handleSpansFlyoutRequest } from '../../requests/traces_request_handler';
import { nanoToMilliSec } from '../common/helper_functions';
import { FlyoutListItem } from './flyout_list_item';

export function SpanDetailFlyout(props: {
  http: HttpSetup;
  spanId: string;
  isFlyoutVisible: boolean;
  closeFlyout: () => void;
  addSpanFilter: (field: string, value: any) => void;
}) {
  const [span, setSpan] = useState<any>({});

  useEffect(() => {
    handleSpansFlyoutRequest(props.http, props.spanId, setSpan);
  }, [props.spanId]);

  const getListItem = (field: string, title: React.ReactNode, description: React.ReactNode) => {
    return (
      <FlyoutListItem
        title={title}
        description={description}
        key={`list-item-${title}`}
        addSpanFilter={() => props.addSpanFilter(field, span[field])}
      />
    );
  };

  const isEmpty = (value) => {
    return (
      value == null ||
      (value.hasOwnProperty('length') && value.length === 0) ||
      (value.constructor === Object && Object.keys(value).length === 0)
    );
  };

  const renderContent = () => {
    if (!span || _.isEmpty(span)) return '-';
    const overviewList = [
      getListItem(
        'spanId',
        'Span ID',
        span.spanId ? (
          <EuiFlexGroup gutterSize="xs" style={{ marginTop: -4, marginBottom: -4 }}>
            <EuiFlexItem grow={false}>
              <EuiCopy textToCopy={span.spanId}>
                {(copy) => (
                  <EuiButtonIcon aria-label="copy-button" onClick={copy} iconType="copyClipboard" />
                )}
              </EuiCopy>
            </EuiFlexItem>
            <EuiFlexItem>{span.spanId}</EuiFlexItem>
          </EuiFlexGroup>
        ) : (
          '-'
        )
      ),
      getListItem(
        'parentSpanId',
        'Parent span ID',
        span.parentSpanId ? (
          <EuiFlexGroup gutterSize="xs" style={{ marginTop: -4, marginBottom: -4 }}>
            <EuiFlexItem grow={false}>
              <EuiCopy textToCopy={span.parentSpanId}>
                {(copy) => (
                  <EuiButtonIcon aria-label="copy-button" onClick={copy} iconType="copyClipboard" />
                )}
              </EuiCopy>
            </EuiFlexItem>
            <EuiFlexItem>{span.parentSpanId}</EuiFlexItem>
          </EuiFlexGroup>
        ) : (
          '-'
        )
      ),
      getListItem('serviceName', 'Service', span.serviceName || '-'),
      getListItem('name', 'Operation', span.name || '-'),
      getListItem(
        'durationInNanos',
        'Duration',
        `${_.round(nanoToMilliSec(Math.max(0, span.durationInNanos)), 2)} ms`
      ),
      getListItem(
        'startTime',
        'Start time',
        moment(span.startTime).format(TRACE_ANALYTICS_DATE_FORMAT)
      ),
      getListItem('endTime', 'End time', moment(span.endTime).format(TRACE_ANALYTICS_DATE_FORMAT)),
      getListItem(
        'status.code',
        'Errors',
        span['status.code'] === 2 ? (
          <EuiText color="danger" size="s" style={{fontWeight: 700}}>
            Yes
          </EuiText>
        ) : (
          'No'
        )
      ),
    ];
    const ignoredKeys = new Set([
      'spanId',
      'parentSpanId',
      'serviceName',
      'name',
      'durationInNanos',
      'startTime',
      'endTime',
      'status.code',
      'events',
      'traceId',
      'traceGroup',
      'traceGroupFields.endTime',
      'traceGroupFields.statusCode',
      'traceGroupFields.durationInNanos',
    ]);
    const attributesList = Object.keys(span)
      .filter((key) => !ignoredKeys.has(key))
      .sort((keyA, keyB) => {
        const isANull = isEmpty(span[keyA]);
        const isBNull = isEmpty(span[keyB]);
        if ((isANull && isBNull) || (!isANull && !isBNull)) return keyA < keyB ? -1 : 1;
        if (isANull) return 1;
        return -1;
      })
      .map((key) => {
        if (isEmpty(span[key])) return getListItem(key, key, '-');
        let value = span[key];
        if (typeof value === 'object') value = JSON.stringify(value);
        return getListItem(key, key, value);
      });

    const eventsComponent = _.isEmpty(span['events']) ? null : (
      <>
        <EuiText size="m">
          <span className="panel-title">Event</span>
        </EuiText>
        <EuiCodeBlock language="json" paddingSize="s" isCopyable overflowHeight={400}>
          {JSON.stringify(span['events'], null, 2)}
        </EuiCodeBlock>
        <EuiSpacer size="xs" />
        <EuiHorizontalRule margin="s" />
      </>
    );

    return (
      <>
        <EuiText size="m">
          <span className="panel-title">Overview</span>
        </EuiText>
        <EuiSpacer size="s" />
        {overviewList}
        <EuiSpacer size="xs" />
        <EuiHorizontalRule margin="s" />
        {eventsComponent}
        <EuiText size="m">
          <span className="panel-title">Span attributes</span>
          {attributesList.length === 0 || attributesList.length ? (
            <span className="panel-title-count">{` (${attributesList.length})`}</span>
          ) : null}
        </EuiText>
        <EuiSpacer size="s" />
        {attributesList}
      </>
    );
  };

  return (
    <>
      <EuiFlyout data-test-subj="spanDetailFlyout" onClose={props.closeFlyout} size="s">
        <EuiFlyoutHeader hasBorder>
          <EuiTitle>
            <h2>Span detail</h2>
          </EuiTitle>
        </EuiFlyoutHeader>
        <EuiFlyoutBody>{renderContent()}</EuiFlyoutBody>
      </EuiFlyout>
    </>
  );
}
