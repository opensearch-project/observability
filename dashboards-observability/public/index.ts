/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import './components/trace_analytics/index.scss';
import './components/notebooks/index.scss'

import { PluginInitializer, PluginInitializerContext } from '../../../src/core/public';
import {
  ObservabilityPlugin
} from './plugin';

export { ObservabilityPlugin as Plugin };

export const plugin = (initializerContext: PluginInitializerContext) => new ObservabilityPlugin(initializerContext);