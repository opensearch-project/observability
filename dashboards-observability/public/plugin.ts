/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { AppMountParameters, CoreSetup, CoreStart, Plugin } from '../../../src/core/public';
import {
  observabilityID,
  observabilityPluginOrder,
  observabilityTitle,
} from '../common/constants/shared';
import PPLService from './services/requests/ppl';
import DSLService from './services/requests/dsl';
import TimestampUtils from './services/timestamp/timestamp';
import SavedObjects from './services/saved_objects/event_analytics/saved_objects';
import { AppPluginStartDependencies, ObservabilitySetup, ObservabilityStart } from './types';
import { convertLegacyNotebooksUrl } from './components/notebooks/components/helpers/legacy_route_helpers';
import { convertLegacyTraceAnalyticsUrl } from './components/trace_analytics/components/common/legacy_route_helpers';
import { uiSettingsService } from '../common/utils';
import { QueryManager } from '../common/query_manager';
export class ObservabilityPlugin implements Plugin<ObservabilitySetup, ObservabilityStart> {
  public setup(core: CoreSetup): ObservabilitySetup {
    uiSettingsService.init(core.uiSettings, core.notifications);

    // redirect legacy notebooks URL to current URL under observability
    if (window.location.pathname.includes('notebooks-dashboards')) {
      window.location.assign(convertLegacyNotebooksUrl(window.location));
    }

    // redirect legacy trace analytics URL to current URL under observability
    if (window.location.pathname.includes('trace-analytics-dashboards')) {
      window.location.assign(convertLegacyTraceAnalyticsUrl(window.location));
    }

    core.application.register({
      id: observabilityID,
      title: observabilityTitle,
      category: {
        id: 'opensearch',
        label: 'OpenSearch Plugins',
        order: 2000,
      },
      order: observabilityPluginOrder,
      async mount(params: AppMountParameters) {
        const { Observability } = await import('./components/index');
        const [coreStart, depsStart] = await core.getStartServices();
        const pplService = new PPLService(coreStart.http);
        const dslService = new DSLService(coreStart.http);
        const savedObjects = new SavedObjects(coreStart.http);
        const timestampUtils = new TimestampUtils(dslService, pplService);
        const qm = new QueryManager();
        return Observability(
          coreStart,
          depsStart as AppPluginStartDependencies,
          params,
          pplService,
          dslService,
          savedObjects,
          timestampUtils,
          qm
        );
      },
    });

    // Return methods that should be available to other plugins
    return {};
  }
  public start(core: CoreStart): ObservabilityStart {
    return {};
  }
  public stop() {}
}
