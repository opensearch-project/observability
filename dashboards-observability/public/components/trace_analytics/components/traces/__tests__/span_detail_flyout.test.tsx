/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { waitFor } from '@testing-library/react';
import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import ReactDOM from 'react-dom';
import { act } from 'react-dom/test-utils';
import { HttpResponse } from '../../../../../../../../src/core/public';
import { TEST_SPAN_RESPONSE } from '../../../../../../test/constants';
import httpClientMock from '../../../../../../test/__mocks__/httpClientMock';
import { SpanDetailFlyout } from '../span_detail_flyout';

describe('<SpanDetailFlyout /> spec', () => {
  configure({ adapter: new Adapter() });

  it('renders the empty component', async () => {
    httpClientMock.post = jest.fn(() =>
      Promise.resolve(({ hits: { hits: [], total: { value: 0 } } } as unknown) as HttpResponse)
    );
    const closeFlyout = jest.fn();
    const addSpanFilter = jest.fn();
    const utils = await mount(
      <SpanDetailFlyout
        http={httpClientMock}
        spanId="test"
        isFlyoutVisible={true}
        closeFlyout={closeFlyout}
        addSpanFilter={addSpanFilter}
        mode="data_prepper"
      />
    );
    utils.update();
    await waitFor(() => {
      expect(utils).toMatchSnapshot();
    });
  });

  it('renders the component with data', async () => {
    httpClientMock.post = jest.fn(() =>
      Promise.resolve((TEST_SPAN_RESPONSE as unknown) as HttpResponse)
    );
    let container = document.createElement('div');
    const closeFlyout = jest.fn();
    const addSpanFilter = jest.fn();
    await act(() => {
      ReactDOM.render(
        <SpanDetailFlyout
          http={httpClientMock}
          spanId="test"
          isFlyoutVisible={true}
          closeFlyout={closeFlyout}
          addSpanFilter={addSpanFilter}
          mode="data_prepper"
        />,
        container
      );
    });
    expect(container).toMatchSnapshot();
  });
});
