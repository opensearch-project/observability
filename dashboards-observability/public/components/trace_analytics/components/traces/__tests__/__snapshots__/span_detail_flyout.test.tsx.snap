// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SpanDetailFlyout /> spec renders the component with data 1`] = `<div />`;

exports[`<SpanDetailFlyout /> spec renders the empty component 1`] = `
<SpanDetailFlyout
  addSpanFilter={[MockFunction]}
  closeFlyout={[MockFunction]}
  http={[MockFunction]}
  isFlyoutVisible={true}
  spanId="test"
>
  <EuiFlyout
    data-test-subj="spanDetailFlyout"
    onClose={[MockFunction]}
    size="s"
  >
    <EuiWindowEvent
      event="keydown"
      handler={[Function]}
    />
    <EuiOverlayMask
      headerZindexLocation="below"
      onClick={[MockFunction]}
    >
      <Portal
        containerInfo={
          <div
            class="euiOverlayMask euiOverlayMask--belowHeader"
          >
            <div
              aria-hidden="true"
              data-aria-hidden="true"
              data-focus-guard="true"
              style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
              tabindex="0"
            />
            <div
              data-focus-lock-disabled="false"
            >
              <div
                class="euiFlyout euiFlyout--small euiFlyout--paddingLarge"
                data-test-subj="spanDetailFlyout"
                role="dialog"
                tabindex="-1"
              >
                <button
                  aria-label="Close this dialog"
                  class="euiButtonIcon euiButtonIcon--text euiButtonIcon--empty euiButtonIcon--xSmall euiFlyout__closeButton euiFlyout__closeButton--inside"
                  data-test-subj="euiFlyoutCloseButton"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="euiIcon euiIcon--medium euiIcon--inherit euiIcon-isLoading euiButtonIcon__icon"
                    focusable="false"
                    height="16"
                    role="img"
                    viewBox="0 0 16 16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  />
                </button>
                <div
                  class="euiFlyoutHeader euiFlyoutHeader--hasBorder"
                >
                  <h2
                    class="euiTitle euiTitle--medium"
                  >
                    Span detail
                  </h2>
                </div>
                <div
                  class="euiFlyoutBody"
                >
                  <div
                    class="euiFlyoutBody__overflow"
                    tabindex="0"
                  >
                    <div
                      class="euiFlyoutBody__overflowContent"
                    >
                      -
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              aria-hidden="true"
              data-aria-hidden="true"
              data-focus-guard="true"
              style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
              tabindex="0"
            />
          </div>
        }
      >
        <EuiFocusTrap
          clickOutsideDisables={false}
          disabled={false}
          noIsolation={true}
          returnFocus={true}
          scrollLock={false}
        >
          <ForwardRef
            enabled={true}
            noIsolation={true}
            onClickOutside={[Function]}
            returnFocus={true}
            scrollLock={false}
          >
            <ForwardRef
              enabled={true}
              noIsolation={true}
              onClickOutside={[Function]}
              returnFocus={true}
              scrollLock={false}
              sideCar={[Function]}
            >
              <ForwardRef(FocusLockUI)
                as={
                  Object {
                    "$$typeof": Symbol(react.forward_ref),
                    "classNames": Object {
                      "fullWidth": "width-before-scroll-bar",
                      "zeroRight": "right-scroll-bar-position",
                    },
                    "defaultProps": Object {
                      "enabled": true,
                      "inert": false,
                      "removeScrollBar": true,
                    },
                    "render": [Function],
                  }
                }
                autoFocus={true}
                crossFrame={true}
                disabled={false}
                lockProps={
                  Object {
                    "allowPinchZoom": undefined,
                    "as": undefined,
                    "enabled": false,
                    "inert": undefined,
                    "onMouseDown": [Function],
                    "onTouchStart": [Function],
                    "shards": undefined,
                    "sideCar": [Function],
                    "style": undefined,
                  }
                }
                noFocusGuards={false}
                onActivation={[Function]}
                onDeactivation={[Function]}
                persistentFocus={false}
                returnFocus={true}
                sideCar={[Function]}
              >
                <div
                  data-focus-guard={true}
                  key="guard-first"
                  style={
                    Object {
                      "height": "0px",
                      "left": "1px",
                      "overflow": "hidden",
                      "padding": 0,
                      "position": "fixed",
                      "top": "1px",
                      "width": "1px",
                    }
                  }
                  tabIndex={0}
                />
                <RequireSideCar
                  autoFocus={true}
                  crossFrame={true}
                  disabled={false}
                  id={Object {}}
                  observed={
                    <div
                      data-focus-lock-disabled="false"
                    >
                      <div
                        class="euiFlyout euiFlyout--small euiFlyout--paddingLarge"
                        data-test-subj="spanDetailFlyout"
                        role="dialog"
                        tabindex="-1"
                      >
                        <button
                          aria-label="Close this dialog"
                          class="euiButtonIcon euiButtonIcon--text euiButtonIcon--empty euiButtonIcon--xSmall euiFlyout__closeButton euiFlyout__closeButton--inside"
                          data-test-subj="euiFlyoutCloseButton"
                          type="button"
                        >
                          <svg
                            aria-hidden="true"
                            class="euiIcon euiIcon--medium euiIcon--inherit euiIcon-isLoading euiButtonIcon__icon"
                            focusable="false"
                            height="16"
                            role="img"
                            viewBox="0 0 16 16"
                            width="16"
                            xmlns="http://www.w3.org/2000/svg"
                          />
                        </button>
                        <div
                          class="euiFlyoutHeader euiFlyoutHeader--hasBorder"
                        >
                          <h2
                            class="euiTitle euiTitle--medium"
                          >
                            Span detail
                          </h2>
                        </div>
                        <div
                          class="euiFlyoutBody"
                        >
                          <div
                            class="euiFlyoutBody__overflow"
                            tabindex="0"
                          >
                            <div
                              class="euiFlyoutBody__overflowContent"
                            >
                              -
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  }
                  onActivation={[Function]}
                  onDeactivation={[Function]}
                  persistentFocus={false}
                  returnFocus={[Function]}
                  shards={Array []}
                  sideCar={
                    Object {
                      "assignMedium": [Function],
                      "assignSyncMedium": [Function],
                      "options": Object {
                        "async": true,
                        "ssr": false,
                      },
                      "read": [Function],
                      "useMedium": [Function],
                    }
                  }
                >
                  <SideCar
                    autoFocus={true}
                    crossFrame={true}
                    disabled={false}
                    id={Object {}}
                    observed={
                      <div
                        data-focus-lock-disabled="false"
                      >
                        <div
                          class="euiFlyout euiFlyout--small euiFlyout--paddingLarge"
                          data-test-subj="spanDetailFlyout"
                          role="dialog"
                          tabindex="-1"
                        >
                          <button
                            aria-label="Close this dialog"
                            class="euiButtonIcon euiButtonIcon--text euiButtonIcon--empty euiButtonIcon--xSmall euiFlyout__closeButton euiFlyout__closeButton--inside"
                            data-test-subj="euiFlyoutCloseButton"
                            type="button"
                          >
                            <svg
                              aria-hidden="true"
                              class="euiIcon euiIcon--medium euiIcon--inherit euiIcon-isLoading euiButtonIcon__icon"
                              focusable="false"
                              height="16"
                              role="img"
                              viewBox="0 0 16 16"
                              width="16"
                              xmlns="http://www.w3.org/2000/svg"
                            />
                          </button>
                          <div
                            class="euiFlyoutHeader euiFlyoutHeader--hasBorder"
                          >
                            <h2
                              class="euiTitle euiTitle--medium"
                            >
                              Span detail
                            </h2>
                          </div>
                          <div
                            class="euiFlyoutBody"
                          >
                            <div
                              class="euiFlyoutBody__overflow"
                              tabindex="0"
                            >
                              <div
                                class="euiFlyoutBody__overflowContent"
                              >
                                -
                              </div>
                            </div>
                          </div>
                        </div>
                      </div>
                    }
                    onActivation={[Function]}
                    onDeactivation={[Function]}
                    persistentFocus={false}
                    returnFocus={[Function]}
                    shards={Array []}
                    sideCar={
                      Object {
                        "assignMedium": [Function],
                        "assignSyncMedium": [Function],
                        "options": Object {
                          "async": true,
                          "ssr": false,
                        },
                        "read": [Function],
                        "useMedium": [Function],
                      }
                    }
                  >
                    <SideEffect(FocusWatcher)
                      autoFocus={true}
                      crossFrame={true}
                      disabled={false}
                      id={Object {}}
                      observed={
                        <div
                          data-focus-lock-disabled="false"
                        >
                          <div
                            class="euiFlyout euiFlyout--small euiFlyout--paddingLarge"
                            data-test-subj="spanDetailFlyout"
                            role="dialog"
                            tabindex="-1"
                          >
                            <button
                              aria-label="Close this dialog"
                              class="euiButtonIcon euiButtonIcon--text euiButtonIcon--empty euiButtonIcon--xSmall euiFlyout__closeButton euiFlyout__closeButton--inside"
                              data-test-subj="euiFlyoutCloseButton"
                              type="button"
                            >
                              <svg
                                aria-hidden="true"
                                class="euiIcon euiIcon--medium euiIcon--inherit euiIcon-isLoading euiButtonIcon__icon"
                                focusable="false"
                                height="16"
                                role="img"
                                viewBox="0 0 16 16"
                                width="16"
                                xmlns="http://www.w3.org/2000/svg"
                              />
                            </button>
                            <div
                              class="euiFlyoutHeader euiFlyoutHeader--hasBorder"
                            >
                              <h2
                                class="euiTitle euiTitle--medium"
                              >
                                Span detail
                              </h2>
                            </div>
                            <div
                              class="euiFlyoutBody"
                            >
                              <div
                                class="euiFlyoutBody__overflow"
                                tabindex="0"
                              >
                                <div
                                  class="euiFlyoutBody__overflowContent"
                                >
                                  -
                                </div>
                              </div>
                            </div>
                          </div>
                        </div>
                      }
                      onActivation={[Function]}
                      onDeactivation={[Function]}
                      persistentFocus={false}
                      returnFocus={[Function]}
                      shards={Array []}
                    >
                      <FocusWatcher
                        autoFocus={true}
                        crossFrame={true}
                        disabled={false}
                        id={Object {}}
                        observed={
                          <div
                            data-focus-lock-disabled="false"
                          >
                            <div
                              class="euiFlyout euiFlyout--small euiFlyout--paddingLarge"
                              data-test-subj="spanDetailFlyout"
                              role="dialog"
                              tabindex="-1"
                            >
                              <button
                                aria-label="Close this dialog"
                                class="euiButtonIcon euiButtonIcon--text euiButtonIcon--empty euiButtonIcon--xSmall euiFlyout__closeButton euiFlyout__closeButton--inside"
                                data-test-subj="euiFlyoutCloseButton"
                                type="button"
                              >
                                <svg
                                  aria-hidden="true"
                                  class="euiIcon euiIcon--medium euiIcon--inherit euiIcon-isLoading euiButtonIcon__icon"
                                  focusable="false"
                                  height="16"
                                  role="img"
                                  viewBox="0 0 16 16"
                                  width="16"
                                  xmlns="http://www.w3.org/2000/svg"
                                />
                              </button>
                              <div
                                class="euiFlyoutHeader euiFlyoutHeader--hasBorder"
                              >
                                <h2
                                  class="euiTitle euiTitle--medium"
                                >
                                  Span detail
                                </h2>
                              </div>
                              <div
                                class="euiFlyoutBody"
                              >
                                <div
                                  class="euiFlyoutBody__overflow"
                                  tabindex="0"
                                >
                                  <div
                                    class="euiFlyoutBody__overflowContent"
                                  >
                                    -
                                  </div>
                                </div>
                              </div>
                            </div>
                          </div>
                        }
                        onActivation={[Function]}
                        onDeactivation={[Function]}
                        persistentFocus={false}
                        returnFocus={[Function]}
                        shards={Array []}
                      />
                    </SideEffect(FocusWatcher)>
                  </SideCar>
                </RequireSideCar>
                <ForwardRef
                  data-focus-lock-disabled={false}
                  enabled={false}
                  inert={false}
                  onBlur={[Function]}
                  onFocus={[Function]}
                  onMouseDown={[Function]}
                  onTouchStart={[Function]}
                  removeScrollBar={true}
                  sideCar={[Function]}
                >
                  <div
                    data-focus-lock-disabled={false}
                    onBlur={[Function]}
                    onFocus={[Function]}
                    onMouseDown={[Function]}
                    onScrollCapture={[Function]}
                    onTouchMoveCapture={[Function]}
                    onTouchStart={[Function]}
                    onWheelCapture={[Function]}
                  >
                    <EuiOutsideClickDetector
                      isDisabled={true}
                      onOutsideClick={[Function]}
                    >
                      <div
                        className="euiFlyout euiFlyout--small euiFlyout--paddingLarge"
                        data-test-subj="spanDetailFlyout"
                        onMouseDown={[Function]}
                        onMouseUp={[Function]}
                        onTouchEnd={[Function]}
                        onTouchStart={[Function]}
                        role="dialog"
                        tabIndex={-1}
                      >
                        <EuiI18n
                          default="Close this dialog"
                          token="euiFlyout.closeAriaLabel"
                        >
                          <EuiButtonIcon
                            aria-label="Close this dialog"
                            className="euiFlyout__closeButton euiFlyout__closeButton--inside"
                            color="text"
                            data-test-subj="euiFlyoutCloseButton"
                            display="empty"
                            iconType="cross"
                            onClick={[Function]}
                          >
                            <button
                              aria-label="Close this dialog"
                              className="euiButtonIcon euiButtonIcon--text euiButtonIcon--empty euiButtonIcon--xSmall euiFlyout__closeButton euiFlyout__closeButton--inside"
                              data-test-subj="euiFlyoutCloseButton"
                              disabled={false}
                              onClick={[Function]}
                              type="button"
                            >
                              <EuiIcon
                                aria-hidden="true"
                                className="euiButtonIcon__icon"
                                color="inherit"
                                size="m"
                                type="cross"
                              >
                                <EuiIconEmpty
                                  aria-hidden={true}
                                  className="euiIcon euiIcon--medium euiIcon--inherit euiIcon-isLoading euiButtonIcon__icon"
                                  focusable="false"
                                  role="img"
                                  style={null}
                                >
                                  <svg
                                    aria-hidden={true}
                                    className="euiIcon euiIcon--medium euiIcon--inherit euiIcon-isLoading euiButtonIcon__icon"
                                    focusable="false"
                                    height={16}
                                    role="img"
                                    style={null}
                                    viewBox="0 0 16 16"
                                    width={16}
                                    xmlns="http://www.w3.org/2000/svg"
                                  />
                                </EuiIconEmpty>
                              </EuiIcon>
                            </button>
                          </EuiButtonIcon>
                        </EuiI18n>
                        <EuiFlyoutHeader
                          hasBorder={true}
                        >
                          <div
                            className="euiFlyoutHeader euiFlyoutHeader--hasBorder"
                          >
                            <EuiTitle>
                              <h2
                                className="euiTitle euiTitle--medium"
                              >
                                Span detail
                              </h2>
                            </EuiTitle>
                          </div>
                        </EuiFlyoutHeader>
                        <EuiFlyoutBody>
                          <div
                            className="euiFlyoutBody"
                          >
                            <div
                              className="euiFlyoutBody__overflow"
                              tabIndex={0}
                            >
                              <div
                                className="euiFlyoutBody__overflowContent"
                              >
                                -
                              </div>
                            </div>
                          </div>
                        </EuiFlyoutBody>
                      </div>
                    </EuiOutsideClickDetector>
                  </div>
                </ForwardRef>
                <div
                  data-focus-guard={true}
                  style={
                    Object {
                      "height": "0px",
                      "left": "1px",
                      "overflow": "hidden",
                      "padding": 0,
                      "position": "fixed",
                      "top": "1px",
                      "width": "1px",
                    }
                  }
                  tabIndex={0}
                />
              </ForwardRef(FocusLockUI)>
              <RequireSideCar
                noIsolation={true}
                onClickOutside={[Function]}
                setLockProps={[Function]}
                sideCar={
                  Object {
                    "assignMedium": [Function],
                    "assignSyncMedium": [Function],
                    "options": Object {
                      "async": true,
                      "ssr": false,
                    },
                    "read": [Function],
                    "useMedium": [Function],
                  }
                }
              >
                <SideCar
                  noIsolation={true}
                  onClickOutside={[Function]}
                  setLockProps={[Function]}
                  sideCar={
                    Object {
                      "assignMedium": [Function],
                      "assignSyncMedium": [Function],
                      "options": Object {
                        "async": true,
                        "ssr": false,
                      },
                      "read": [Function],
                      "useMedium": [Function],
                    }
                  }
                >
                  <Effect
                    noIsolation={true}
                    onClickOutside={[Function]}
                    setLockProps={[Function]}
                  >
                    <Component>
                      <Sheet
                        styles="
 [data-focus-on-hidden] {
   pointer-events: none !important;
 }
"
                      />
                    </Component>
                  </Effect>
                </SideCar>
              </RequireSideCar>
            </ForwardRef>
          </ForwardRef>
        </EuiFocusTrap>
      </Portal>
    </EuiOverlayMask>
  </EuiFlyout>
</SpanDetailFlyout>
`;
